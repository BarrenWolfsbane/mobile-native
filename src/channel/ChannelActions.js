import React, {
  Component
} from 'react';

import {
  Text,
  Image,
  View,
  ActivityIndicator,
  Button,
  StyleSheet
} from 'react-native';

import {
  observer,
  inject
} from 'mobx-react/native'

import Icon from 'react-native-vector-icons/Ionicons';
import channelService from './ChannelService';
import { MINDS_URI } from '../config/Config';
import ActionSheet from 'react-native-actionsheet';

/**
 * Channel Actions
 */
const title = 'Actions'

@observer
export default class ChannelActions extends Component {

  constructor(props) {
    super(props)
    this.state = {
      selected: '',
      options: this.getOptions(),
    }

    this.handleSelection = this.handleSelection.bind(this);
  }

  showActionSheet() {
    this.state = {
      options: this.getOptions(),
    }
    this.ActionSheet.show();
  }

  handleSelection(i) {
    this.makeAction(this.state.options[i]);
  }

  getOptions() {
    let options = [ 'Cancel' ];
    if(!!this.props.channel.channel.subscribed){
      options.push( 'Unsubscribe' );
    }

    if(!this.props.channel.channel.blocked){
      options.push( 'Block' );
    }

    if(!!this.props.channel.channel.blocked){
      options.push( 'Un-Block' );
    }

    options.push( 'Report' );

    return options;

  }

  makeAction(option) {
    switch (option) {
      case 'Unsubscribe':
        this.props.channel.subscribe().then( (result) => {
          this.setState({
            options: this.getOptions(),
          })
        });
        break;
      case 'Block':
        this.props.channel.toggleBlock().then( (result) => {
          this.setState({
            options: this.getOptions(),
          })
        });
        break;
      case 'Un-Block':
        this.props.channel.toggleBlock().then( (result) => {
          this.setState({
            options: this.getOptions(),
          })
        });
        break;
    }

    
  }
  
  /**
   * Render Header
   */
  render() {

    const channel = this.props.channel.channel;

    return (
      <View>
        <Icon name="md-settings" onPress={() => this.showActionSheet()} size={20} />
        <ActionSheet
          ref={o => this.ActionSheet = o}
          title={title}
          options={this.state.options}
          onPress={this.handleSelection}
          cancelButtonIndex={0}
        />
      </View>
    )
  }
}

const styles = StyleSheet.create({
	wrapper: {
    backgroundColor: '#FFF',
    flex: 1,
  },
  iconclose: {
    flex:1,
  }
});