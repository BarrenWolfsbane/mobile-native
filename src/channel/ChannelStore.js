import {
  observable,
  action
} from 'mobx';

import channelService from './ChannelService';

/**
 * Channel Store
 */
class ChannelStore {
  @observable channel = {};

  @action
  setChannel(ch) {
    this.channel = ch;
  }

  @action
  clearChannel() {
    this.channel = {};
  }

  @action
  load(guid) {
    this.channel = {};
    channelService.load(guid)
      .then(action(response => {
        console.log(response);
        this.channel = response.channel;
      }))
      .catch(err => {
        console.log('error', err);
      });
  }
}

export default new ChannelStore();
