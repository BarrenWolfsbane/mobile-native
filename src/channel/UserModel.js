import { observable, action, runInAction } from 'mobx';
import { MINDS_CDN_URI, GOOGLE_PLAY_STORE } from '../config/Config';
import api from '../common/services/api.service';
import BaseModel from '../common/BaseModel';
import ChannelService from './ChannelService';
import sessionService from '../common/services/session.service';


/**
 * User model
 */
export default class UserModel extends BaseModel {

  /**
   * @var boolean
   */
  @observable blocked;

  /**
   * @var integer
   */
  @observable subscribers_count;

  /**
   * @var integer
   */
  @observable impressions;

  /**
   * @var boolean
   */
  @observable subscribed;
  /**
   * @var boolean
   */
  @observable mature_visibility = false;

  getOwnerIcontime() {
    if (sessionService.getUser().guid === this.guid) {
      return sessionService.getUser().icontime;
    } else {
      return this.icontime;
    }
  }

  @action
  toggleMatureVisibility() {
    if (GOOGLE_PLAY_STORE) return;
    this.mature_visibility = !this.mature_visibility;
  }

  @action
  async toggleSubscription() {
    const value = !this.subscribed;
    this.subscribed = value;
    try {
      const metadata = this.getClientMetadata();
      await ChannelService.toggleSubscription(this.guid, value, metadata);
    } catch (err) {
      runInAction(() => {
        this.subscribed = !value;
      });
      throw err;
    }
  }

  /**
   * current user is owner of the channel
   */
  isOwner() {
    return sessionService.getUser().guid === this.guid;
  }

  /**
   * Get banner source
   * @param {string} size
   */
  getBannerSource(size='medium') {

    if (this.carousels) {
      return {
        uri: this.carousels[0].src
      };
    }
    return { uri: `${MINDS_CDN_URI}fs/v1/banners/${this.guid}/fat/${this.icontime}`, headers: api.buildHeaders()};
  }

  /**
   * Get avatar source
   * @param {string} size
   */
  getAvatarSource(size='medium') {
    return { uri: `${MINDS_CDN_URI}icon/${this.guid}/${size}/${this.getOwnerIcontime()}`, headers: api.buildHeaders()};
  }

  /**
   * Has banner
   */
  hasBanner() {
    return !!this.carousels;
  }
}