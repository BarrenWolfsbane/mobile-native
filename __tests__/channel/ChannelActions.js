import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import ChannelActions from '../../src/channel/ChannelActions';
import UserModel from '../../src/channel/UserModel';
import userFaker from '../../__mocks__/fake/channel/UserFactory'
import ChannelStore from '../../src/channel/ChannelStore';
import features from '../../src/common/services/features.service';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

jest.mock('../../src/channel/ChannelStore');
jest.mock('../../src/channel/UserModel');
jest.mock('../../src/common/services/features.service');
jest.mock('../../AppStores');

/**
 * Tests
 */
describe('channel actions component', () => {
  beforeEach(() => {
    store = new ChannelStore();
    store.channel = new UserModel(userFaker(1));
  });

  it('should renders correctly', () => {

    // return true to has crypto
    features.has.mockReturnValue(true);

    const component = renderer.create(
      <ChannelActions channel={store} />
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  it('should show the correct options', () => {

    const wrapper = shallow(
      <ChannelActions channel={store} />
    );

    let opt = wrapper.instance().getOptions();

    expect(opt).toEqual([ 'Cancel', 'Block', 'Report' ]);

    // if subscribed
    store.channel.subscribed = true;
    opt = wrapper.instance().getOptions();

    expect(opt).toEqual([ 'Cancel', 'Unsubscribe', 'Block', 'Report' ]);

    // if blocked
    store.channel.blocked = true;
    opt = wrapper.instance().getOptions();
    expect(opt).toEqual([ 'Cancel', 'Unsubscribe', 'Un-Block', 'Report' ]);
  });

  it('should show run the correct option', () => {

    const navigation = {navigate: jest.fn()};
    const wrapper = shallow(
      <ChannelActions channel={store} navigation={navigation}/>
    );

    store.channel.subscribed = true;
    store.channel.blocked = true;

    opt = wrapper.instance().makeAction(1);
    expect(store.subscribe).toBeCalled();

    opt = wrapper.instance().makeAction(2);
    expect(store.toggleBlock).toBeCalled();

    opt = wrapper.instance().makeAction(3);
    expect(navigation.navigate).toBeCalled();
  });

});