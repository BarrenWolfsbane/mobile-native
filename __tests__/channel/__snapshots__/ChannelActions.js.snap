// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channel actions component should renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFF",
      "flexDirection": "row",
      "height": 40,
      "justifyContent": "flex-end",
      "paddingLeft": 8,
      "width": 40,
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "center",
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "rgb(70, 144, 214)",
            "fontSize": 40,
          },
          undefined,
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <Text
    allowFontScaling={false}
    onPress={[Function]}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 24,
        },
        Object {
          "color": "#888888",
          "paddingLeft": 10,
        },
        Object {
          "fontFamily": "Ionicons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;
